### DISCUSSION TITLE
11.x TinyMCE: add plugins

### DISCUSSION THREAD

**Discussion by arrstemmler** [pimcore/pimcore#14895] [Q&A / Support]
Hi,
i would use the searchreplace and anchor plugin in a project that we develop on pimcore 11.x.
How can i add the needed plugins?

Hint:
The default plugins are included in [vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js](https://github.com/pimcore/pimcore/blob/230610923ee36ab370261fd13a84619be0162ad8/bundles/TinymceBundle/public/assets/tinymce.js)

**Comment by dvesh3** [comment]
@robertSt7 can we support this somehow?

**Reply by robertSt7** [reply]
Not at the moment. But you can create your own TinyMCE bundle register it and then you have the full control of it

**Reply by leuchtdiode** [reply]
Hi,

I need the anchor plugin as well. Is there no way to add this to a config for TinyMCE?

If we create our own bundle and register it: Do we need to clone the TinyMCE bundle from pimcore to have all the functionalities? This would be overkill and much nicer if we could add plugins via a config.

BR,
Alex

**Reply by cngJo** [reply]
@leuchtdiode,

I solved this in the end by "monkey-patching" the change into [`pimcore/bundles/TinymceBundle/public/js/editor.js`](https://github.com/pimcore/pimcore/blob/11.x/bundles/TinymceBundle/public/js/editor.js) using [`cweagans/composer-patches`](https://github.com/cweagans/composer-patches).

This in the end seems to be the lesser evil than hard-forking the bundle. This way I at least get an error during `composer install` when sth. changed after an update, and I don't need to manually update the fork all the times.

**Reply by leuchtdiode** [reply]
@cngJo Thanks for the info. I think we will go this way as well for now.

Although I think that a pimcore solution via configuration would be nice. Or maybe just add "anchor" into the defaults @dvesh3 @robertSt7? CKEditor had it as well out-of-the-box before pimcore 11. So a lot of people will stumble upon I'm afraid.

BR,
Alex

**Comment by leuchtdiode** [comment]
@cngJo As I saw that you needed the "anchor" plugin as well. Would it be possible to provide us the class for monkey patching with the anchor plugin in it?
Then we would not have to dig into the TinyMCE stuff. That would be awesome :)

BR,
Alex

**Comment by Daniel-Ateles** [comment]
I have the same problem. Need to make the wysiwyg exactly as the old ckeditor so I need these:
searchreplace
advlist
charmap
hr
textcolor
visualblocks

@cngJo could you show us your patch?

**Reply by Daniel-Ateles** [reply]
Aha, it's this list you patched in vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js right @cngJo ?
import * as tinymce from "tinymce";

import 'tinymce/models/dom';
import 'tinymce/icons/default';

// A theme is also required
import 'tinymce/themes/silver';

// Any plugins you want to use has to be imported
import 'tinymce/plugins/link';
import 'tinymce/plugins/table';
import 'tinymce/plugins/wordcount';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/image';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/help';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/code';

**Reply by cngJo** [reply]
Unfortunately, I'm not able to provide the exact patch, since it's part of a proprietary project. 

But in [L90-#93 of bundles/TinymceBundle/public/js/editor.js](https://github.com/pimcore/pimcore/blob/11.x/bundles/TinymceBundle/public/js/editor.js#L90-L93) you find the list of plugins pre-installed by pimcore and in my case, I was just able to add them to this list to get them working.

Also, during the implementation originally, I dug through the source code and found that there is a way to extend this config with the `pimcore_wysiwyg` editable, by passing an object to it. This was not reasonable for use, since we use ~20 `pimcore_wysiwyg` editables across the system and I didn't want to override the entire config for this all the times. 
But it might be reasonable for you, so feel free to explore deeper there. 

CC @Daniel-Ateles & @leuchtdiode, since you've both asked this.

**Reply by Daniel-Ateles** [reply]
Here is the patch. Since it's patch of a vendor no code of your project is shown?
Anyways. It's not enough since Pimcores Tinymcebundle does not contain these plugins so you will get console errors anyways.
So now I guess I have to resort to overriding the Tinymcebundle with our own. 
Such a hassle for so little :(

diff --git a/vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js b/vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js
--- a/vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js
+++ b/vendor/pimcore/pimcore/bundles/TinymceBundle/public/assets/tinymce.js	(date 1707747055476)
@@ -16,4 +16,10 @@
 import 'tinymce/plugins/media';
 import 'tinymce/plugins/help';
 import 'tinymce/plugins/lists';
-import 'tinymce/plugins/code';
\ No newline at end of file
+import 'tinymce/plugins/code';
+
+import 'tinymce/plugins/searchreplace';
+import 'tinymce/plugins/advlist';
+import 'tinymce/plugins/charmap';
+import 'tinymce/plugins/hr';
+import 'tinymce/plugins/visualblocks';

**Comment by leuchtdiode** [comment]
Oh, that's all. Than it looks quite simple. We will give it a try.

Thank you both :)

BR,
Alex

**Reply by Daniel-Ateles** [reply]
It's actually not that simple since the plugins needs to be added to the plugins-directory in Tinymcebundle, and cweagans 1.x can't add files.
I'm banging my head in the wall with this.

**Reply by leuchtdiode** [reply]
Actually I do not even understand where the default plugin files (link, table, etc.) listed in assets/tinymce.js are located? I have the package.json in TinyMceBundle, but no node_modules folder? When/how is webpack via Encore even executed to build the JS files 🤔

**Reply by Daniel-Ateles** [reply]
It seems tinymce loads them "on the fly" as separate files, so all one needs to do is put them in vendor/pimcore/pimcore/bundles/TinymceBundle/public/build/tinymce/plugins
Download them from here:  https://cdnjs.com/libraries/tinymce/6.7.2 (latest version in Pimcore 11.1.5)
Tell me how if manage to copy them there with composer :)

**Reply by leuchtdiode** [reply]
Never tried, but found this a few minutes ago: https://github.com/markocupic/composer-file-copier-plugin

**Reply by Daniel-Ateles** [reply]
Hey @leuchtdiode I managed to make a patch that works that adds a couple of plugins it works, I get searchreplace, hr, advlist and visualblocks. I didn't manage to make such a patch myself, but when I did a PR to pimcore it seems to have worked out fine:
https://patch-diff.githubusercontent.com/raw/pimcore/pimcore/pull/16635.patch

**Reply by leuchtdiode** [reply]
@Daniel-Ateles Thank you. We are fighting with it for 3 hours now. We managed to get anchor in per composer (but not the copy part as my library I referenced last week does not work with the projects composer.json).
But now as we have the anchor plugin in TinyMCE, it is not working as expected, as it always strips the "id" from the a-tag it generated by itself. It's very frustrating.

**Reply by Daniel-Ateles** [reply]
@leuchtdiode that sounds like the "sanitizer" is stripping them, html_sanitizer.yaml. Have you added "id" to the a-tag to allow it? Like so:
a: [ 'class', 'id', 'href', 'target', 'title', 'rel' ]

**Reply by leuchtdiode** [reply]
@Daniel-Ateles It got stripped in the editor already, before saving the document where the sanitizer comes into place. When adding `'extended_valid_elements' => 'a[*]'` to the pimcore_wysiwyg config, it looks like that we got it working now :)
